### PR TITLE
[Issue Refund] Add ShippingLineTax support

### PR DESCRIPTION
**Pull request by Ecarrion**
part of #2843 

# Disclaimer 
PR is not that big, ~500 lines are from the CD model, actual changes are around 250 lines 😇 

# Why

While working on making sure the [API request for refund shipping](https://github.com/woocommerce/woocommerce-ios/pull/3048) works correctly I noticed that the `ShippingLine` entity is missing the `taxes` array needed to be able to refund shipping correctly(Refunding to the specific tax line).

This PR parses and stores an array of `ShippingLineTax` inside of the existing `ShippingLine` that exists within an order.

# How
- Added `ShippingLineTax` core data entity and link it inside `ShippingLine` This includes a new model V34.
- Added `ShippingLineTax` networking entity and link it inside `Networking.ShippingLine`
- Added `ShippingLineTax+ReadOnlyConvertible` to be able to from from Storage type to Networking type.
- Added method on `StorageType+Extensions` to fetch a shipping line tax from storage.
- Updated `OrderUpsertUseCase` to correctly store/handle shipping line items and shipping line item taxes.
- Updated sample Jsons to include shipping tax lines data.
 
# Testing Steps.
- Making sure unit tests pass!

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Model/ShippingLine.swift [hunk 1]
```diff
@@ -68,7 +53,8 @@ extension ShippingLine: Comparable {
             lhs.methodTitle == rhs.methodTitle &&
             lhs.methodID == rhs.methodID &&
             lhs.total == rhs.total &&
-            lhs.totalTax == rhs.totalTax
+            lhs.totalTax == rhs.totalTax &&
+            lhs.taxes == rhs.taxes
```
Comment (jaclync) [inline Networking/Networking/Model/ShippingLine.swift]: nit: would it be safer if we remove this manual Equatable implementation and just let `ShippingLine` implement `Equatable` automatically?

File: Networking/Networking/Model/ShippingLineTax.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    ///
+    public let taxID: Int64
+
+    /// Tax subtotal
+    ///
+    public let subtotal: String
+
+    /// Shipping tax amount
+    ///
+    public let total: String
```
Comment (jaclync) [inline Networking/Networking/Model/ShippingLineTax.swift:17]: it's interesting that I cannot find these fields in the [WC API doc](https://woocommerce.github.io/woocommerce-rest-api-docs/#order-taxes-properties) 🤔

File: Storage/Storage/Model/MIGRATIONS.md [hunk 3]
```diff
@@ -2,6 +2,11 @@
 
 This file documents changes in the WCiOS Storage data model. Please explain any changes to the data model as well as any custom migrations.
 
+## Model 34 (Release 5.4.0.0)
+- @ecarrion 2020-10-21
+- Added `ShippingLineTax`  entity.
+- Added  `taxes` relationship to store `ShippingLine`.
```
Comment (jaclync) [inline Storage/Storage/Model/MIGRATIONS.md]: nit:
```suggestion
- Added `ShippingLineTax` entity.
- Added  `taxes` relationship to `ShippingLine`.
```

File: Storage/Storage/Model/ShippingLine+CoreDataProperties.swift [hunk 4]
```diff
@@ -14,5 +14,23 @@ extension ShippingLine {
     @NSManaged public var total: String?
     @NSManaged public var totalTax: String?
     @NSManaged public var order: Order?
+    @NSManaged public var taxes: Set<ShippingLineTax>?
```
Comment (jaclync) [inline Storage/Storage/Model/ShippingLine+CoreDataProperties.swift:17]: just double checking, the taxes don't have to be ordered right?

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift [hunk 5]
```diff
@@ -184,4 +186,29 @@ struct OrdersUpsertUseCase {
             }
         }
     }
+
+    /// Updates, inserts, or prunes the provided StorageOrderItem's taxes using the provided read-only OrderItem
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift]: nit:
```suggestion
    /// Updates, inserts, or prunes the provided StorageShippingLine's taxes using the provided read-only ShippingLine
```

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift [hunk 6]
```diff
@@ -184,4 +186,29 @@ struct OrdersUpsertUseCase {
...
+            if let existingStorageTax = storage.loadShippingLineTax(shippingID: shippingID, taxID: readOnlyTax.taxID) {
+                existingStorageTax.update(with: readOnlyTax)
+            } else {
+                let newStorageTax = storage.insertNewObject(ofType: Storage.ShippingLineTax.self)
+                newStorageTax.update(with: readOnlyTax)
+                storageItem.addToTaxes(newStorageTax)
+            }
+        }
+
+        // Now, remove any objects that exist in storageOrderItem.taxes but not in readOnlyOrderItem.taxes
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift]: nit:

```suggestion
        // Now, remove any objects that exist in storageItem.taxes but not in readOnlyItem.taxes
```

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift [hunk 7]
```diff
@@ -184,4 +186,29 @@ struct OrdersUpsertUseCase {
...
+        storageItem.taxes?.forEach { storageTax in
+            if readOnlyItem.taxes.first(where: { $0.taxID == storageTax.taxID } ) == nil {
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:208]: could more than one `ShippingLineTax` with the same tax ID exist for different shipping lines? if so, we might want to also check on `$0.shipping.shippingID == readOnlyItem.shippingID` to be sure

**Review by jaclync — APPROVED** [review]
I actually didn't know shipping could be taxable, not to mention multiple taxes in a shipping line! 😮 Code LGTM, just some minor comments :shipit:

File: Networking/Networking/Model/ShippingLine.swift (hunk 1 above)
Comment (Ecarrion) [reply Networking/Networking/Model/ShippingLine.swift]: True!

File: Networking/Networking/Model/ShippingLineTax.swift (hunk 2 above)
Comment (Ecarrion) [reply Networking/Networking/Model/ShippingLineTax.swift:17]: Yeah, documentation is not great :-(

File: Storage/Storage/Model/ShippingLine+CoreDataProperties.swift (hunk 4 above)
Comment (Ecarrion) [reply Storage/Storage/Model/ShippingLine+CoreDataProperties.swift:17]: Not at the moment, this is not displayed anywhere but we need them to know what tax id to refund!

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift (hunk 7 above)
Comment (Ecarrion) [reply Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:208]: I think it's possible but:
- We don't support multiple shipping lines at the moment 
- I think those shipping line taxes would be different objects so we don't risk deleting them

**Pull request merged**